Mathlib/Algebra/Category/ModuleCat/Presheaf/Sheafify.lean

```diff
@@ -0,0 +1,320 @@
+  let S := Presheaf.imageSieve α r ⊓ Presheaf.imageSieve α r' ⊓ Presheaf.imageSieve φ m
+  have hS : S ∈ J X.unop := by
+    refine J.intersection_covering (J.intersection_covering ?_ ?_) ?_
+    all_goals apply Presheaf.imageSieve_mem
+  apply A.isSeparated _ _ hS
+  rintro Y f ⟨⟨⟨r₀ : R₀.obj _, hr₀⟩, ⟨r₀' : R₀.obj _, hr₀'⟩⟩, ⟨m₀, hm₀⟩⟩
+  erw [(A.val.map f.op).map_add, map_smul_eq α φ r m f.op r₀ hr₀ m₀ hm₀,
+    map_smul_eq α φ r' m f.op r₀' hr₀' m₀ hm₀,
+    map_smul_eq α φ (r + r') m f.op (r₀ + r₀') (by rw [map_add, map_add, hr₀, hr₀'])
+      m₀ hm₀, add_smul, map_add]
+
+protected lemma mul_smul : smul α φ (r * r') m = smul α φ r (smul α φ r' m) := by
+  let S := Presheaf.imageSieve α r ⊓ Presheaf.imageSieve α r' ⊓ Presheaf.imageSieve φ m
+  have hS : S ∈ J X.unop := by
+    refine J.intersection_covering (J.intersection_covering ?_ ?_) ?_
+    all_goals apply Presheaf.imageSieve_mem
+  apply A.isSeparated _ _ hS
+  rintro Y f ⟨⟨⟨r₀ : R₀.obj _, hr₀⟩, ⟨r₀' : R₀.obj _, hr₀'⟩⟩, ⟨m₀ : M₀.presheaf.obj _, hm₀⟩⟩
+  erw [map_smul_eq α φ (r * r') m f.op (r₀ * r₀')
+    (by rw [map_mul, map_mul, hr₀, hr₀']) m₀ hm₀, mul_smul,
+    map_smul_eq α φ r (smul α φ r' m) f.op r₀ hr₀ (r₀' • m₀)
+      (map_smul_eq α φ r' m f.op r₀' hr₀' m₀ hm₀).symm]
+
+variable (X)
+
+/-- The module structure on the sections of the sheafification of the underlying
+presheaf of abelian groups of a presheaf of modules. -/
+noncomputable def module : Module (R.val.obj X) (A.val.obj X) where
+  smul r m := smul α φ r m
+  one_smul := Sheafify.one_smul α φ
+  zero_smul := Sheafify.zero_smul α φ
+  smul_zero := Sheafify.smul_zero α φ
+  smul_add := Sheafify.smul_add α φ
+  add_smul := Sheafify.add_smul α φ
+  mul_smul := Sheafify.mul_smul α φ
+
+lemma map_smul :
+    A.val.map π (smul α φ r m) = smul α φ (R.val.map π r) (A.val.map π m) := by
+  let S := Presheaf.imageSieve α (R.val.map π r) ⊓ Presheaf.imageSieve φ (A.val.map π m)
+  have hS : S ∈ J Y.unop := by
+    apply J.intersection_covering
+    all_goals apply Presheaf.imageSieve_mem
+  apply A.isSeparated _ _ hS
+  rintro Y f ⟨⟨r₀, hr₀⟩, ⟨m₀, hm₀⟩⟩
+  erw [← comp_apply, ← Functor.map_comp,
+    map_smul_eq α φ r m (π ≫ f.op) r₀ (by rw [hr₀, Functor.map_comp, comp_apply]) m₀
+      (by erw [hm₀, Functor.map_comp, comp_apply]; rfl),
+    map_smul_eq α φ (R.val.map π r) (A.val.map π m) f.op r₀ hr₀ m₀ hm₀]
+
+end Sheafify
+
+/-- Assuming `α : R₀ ⟶ R.val` is the sheafification map of a presheaf of rings `R₀`
+and `φ : M₀.presheaf ⟶ A.val` is the sheafification map of the underlying
+sheaf of abelian groups of a presheaf of modules `M₀` over `R₀`, this is
+the sheaf of modules over `R` which is obtained by endowing the sections of
+`A.val` with a scalar multiplication. -/
+noncomputable def sheafify : SheafOfModules.{v} R where
+  val :=
+    { presheaf := A.val
+      module := Sheafify.module α φ
+      map_smul := fun _ _ _ => by apply Sheafify.map_smul }
+  isSheaf := A.cond
+
```

Comment:
Am I understanding the PR correctly that everything except `sheafify` is considered implementation details and will never be used outside this file? Can you provide an `example` to show how to use `sheafify` if I have a sheaf of rings `R` and a presheaf of modules `M`?

Reply:
In this PR, I only construct the associated sheaf of modules. The more developed API (including the adjunction between presheaves and sheaves) will be in PR #13441. But, I am adding an example.